python/dgl/contrib/pytorch/multi_gpu_dataloader.py

```diff
@@ -0,0 +1,245 @@
+##
+#   Copyright 2021 Contributors
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+import dgl
+from dgl import utils
+from dgl.frame import Frame
+import torch as th
+from ...dataloading import NodeDataLoader
+from ...partition import NDArrayPartition, create_edge_partition_from_nodes
+from ..multi_gpu_datastore import MultiGPUDataStore
+from typing import Mapping
+
+def _load_tensor(tensor, device, comm, part):
+    loaded_tensor = None
+    if comm:
+        loaded_tensor = MultiGPUDataStore(shape=tensor.shape, dtype=tensor.dtype,
+            device=device, comm=comm, partition=part)
+        loaded_tensor.all_set_global(tensor)
+    else:
+        loaded_tensor = tensor.to(device)
+    return loaded_tensor
+
+def _gather_row(tensor, index):
+    if isinstance(tensor, MultiGPUDataStore):
+        return tensor.all_gather_row(index)
+    else:
+        return tensor[index]
+
+class _NodeDataIterator:
+    def __init__(self, it, n_feat, e_feat, node_feat, node_label):
+        self._it = it
+        # {ntype: {feature: tensor}}
+        self._n_feat = n_feat
+        # {etype: {feature: tensor}}
+        self._e_feat = e_feat
+        self._node_feat = node_feat
+        self._node_label = node_label
+
+    # Make this an iterator for PyTorch Lightning compatibility
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        input_nodes, output_nodes, blocks = next(self._it)
+
+        # re-attach node features
+        for block in blocks:
+            node_frames = []
+            # have to handle src and dst nodes separately in a block
+            for ntype in block.srctypes:
+                index = block.srcnodes[ntype].data[dgl.NID]
+                frame = Frame(num_rows=len(index))
+                for k, v in self._n_feat[ntype].items():
+                    data = _gather_row(v, index)
+                    frame.update_column(k, data)
+                node_frames.append(frame)
+            for ntype in block.dsttypes:
```

Comment:
What if an ntype appears in both block.srctypes and block.dsttypes?

Reply:
This is the case for blocks. In the homogeneous case, we'll have a `_N` for src and dst. See #2986.

I essentially took this from https://github.com/dmlc/dgl/blob/master/python/dgl/utils/internal.py#L793. I had originally just iterated `block.ntypes`, but this breaks for homogeneous graphs due to having the duplicated type.